[Bug]: Cron bug TypeError: Cannot read properties of undefined (reading 'runningAtMs')

### Bug type

Regression (worked before, now fails)

### Beta release blocker

Yes

### Summary

Cron 调度器 bug 定位：

TypeError: Cannot read properties of undefined (reading 'runningAtMs')

错误发生在 cron.start() → runMissedJobs() → planStartupCatchup() 执行期间，job 对象在内存中被替换成了缺少 state 的版本。这是 OpenClaw 2026.4.9 的 cron 子系统 bug。

CLI 无法调用的原因：
openclaw cron add/list 通过 WebSocket 连接 gateway RPC 调用 cron 方法，gateway 每次重启后 cron 都会崩溃重启，WebSocket 握手就卡住了，导致 CLI 无响应。
回复 紫月: 
Cron 调度器 bug 定位：

TypeError：无法读取未定义属性（读取“runningAtMs”）

错误发生在 cron.start（） → runMissedJobs（） → planStartupCatchup（） 执行期间，job 对象在内存中被替换成了缺少 state 的版本。这是 OpenClaw 2026.4.9 的 cron 子系统 bug。
升级到docker  4.11
CLI 无法调用的原因：
openclaw cron add/list 通过 WebSocket 连接 gateway RPC 调用 cron 方法，gateway 每次重启后 cron 都会崩溃重启，WebSocket 握手就卡住了，导致 CLI 无响应。  重新检测doker opencalw 4.11版本还有没有这个问题
已确认 OpenClaw 4.11 版本仍然存在这个 bug。
问题定位
文件: /app/dist/server.impl-CsRRyd9F.js，第 6582 行，start() 函数内：

for (const job of jobs) if (typeof job.state.runningAtMs === "number") {
 ^^^^^^^^^^^^^^^ 没有 guard
**问题：**直接访问 job.state.runningAtMs，没有先检查 job.state 是否存在。
对比其他函数
同一文件中的 isRunnableJob 和 isJobDue 都有防御性检查：

// isRunnableJob 第 6211-6212 行
if (!job.state) job.state = {};
if (typeof job.state.runningAtMs === "number") return false;

// isJobDue 第 3809-3810 行  
if (!job.state) job.state = {};

### Steps to reproduce


今日日志 (/tmp/openclaw/openclaw-2026-04-13.log)
15:03:45 | WARN | Config overwrite
15:05:41 | WARN | Gateway 绑定到非 loopback 地址
15:05:41 | WARN | dangerouslyDisableDeviceAuth=true 安全警告
15:05:42 | INFO | cron: started (jobs: 0, nextWakeAtMs: null)
15:06:06 | WARN | pricing bootstrap failed: TimeoutError

Bug 仍存在于 4.11
/app/dist/server.impl-CsRRyd9F.js 第 6582 行：

for (const job of jobs) if (typeof job.state.runningAtMs === "number") {
//                                      ^^^^^^^^^^^^^^^ 没有 guard



对比有防护的地方（第 6211-6212 行）：

if (!job.state) job.state = {};  // ← 有这个 guard
if (typeof job.state.runningAtMs === "number") return false;

### Expected behavior

今日日志 (/tmp/openclaw/openclaw-2026-04-13.log)
<body>
<!--StartFragment--><div data-lark-html-role="root">
时间 | 级别 | 内容
-- | -- | --
15:03:45 | WARN | Config overwrite
15:05:41 | WARN | Gateway 绑定到非 loopback 地址
15:05:41 | WARN | dangerouslyDisableDeviceAuth=true 安全警告
15:05:42 | INFO | cron: started (jobs: 0, nextWakeAtMs: null)
15:06:06 | WARN | pricing bootstrap failed: TimeoutError

</div><!--EndFragment-->
</body>
</html>
cron 当前状态：0 个 jobs，所以 start() 里的 bug 没有触发。
Bug 仍存在于 4.11
/app/dist/server.impl-CsRRyd9F.js 第 6582 行：

for (const job of jobs) if (typeof job.state.runningAtMs === "number") {
//                                      ^^^^^^^^^^^^^^^ 没有 guard



对比有防护的地方（第 6211-6212 行）：

if (!job.state) job.state = {};  // ← 有这个 guard
if (typeof job.state.runningAtMs === "number") return false;



触发条件： store 里存有 job，但某个 job 缺少 state 字段（老版本创建/数据损坏）。

### Actual behavior

回复 紫月: 
检测我的cron任务
检测结果：

当前系统状态：
Cron 任务：0 个（存储文件 /home/node/.openclaw/cron/jobs.json 不存在，尚未初始化）
CLI 状态：卡住 — 正如你所说，gateway 重启后 WebSocket RPC 挂起，CLI 命令无响应
问题确认：
openclaw cron list — CLI 卡住，无法通过 WebSocket 连接
Cron store 文件夹 /home/node/.openclaw/cron/ 不存在
日志显示最后 cron 启动时 jobs 数量为 0


<html>
<body>
<!--StartFragment--><div data-lark-html-role="root">
项目 | 状态
-- | --
Cron 存储路径 | /home/node/.openclaw/cron/jobs.json
当前任务数 | 0
CLI 连接 | 失败 — pairing required (WebSocket RPC 需要配对)
Bug 状态 | start() 第 6582 行仍缺少 if (!job.state) job.state = {} guard

</div><!--EndFragment-->
</body>
</html>

15:05:42 [INFO] cron: started, jobs: 0, nextWakeAtMs: null
15:14:40 [ERROR] gateway connect failed: pairing required (CLI cron list)
15:16:33 [ERROR] gateway connect failed: pairing required (CLI cron list)
CLI 卡住原因： Gateway 绑定 LAN 地址 + dangerouslyDisableDeviceAuth=true + pairing 模式冲突。
 


Cron 任务：0 个 — 存储文件为空或不存在
Bug 仍存在于 4.11 — 未修复
CLI 无法使用 — WebSocket pairing 问题
守护进程脚本位于：

/home/node/.openclaw/workspace/scripts/juya_ai_daily_scheduler.py
推送脚本：
/home/node/.openclaw/workspace/scripts/juya_ai_daily.py

### OpenClaw version

2026.4.11

### Operating system

Debian13

### Install method

Docker

### Model

MinMax2.7

### Provider / routing chain

Opneclaw->Openclaw-gateway->mainmax2.7->feishu

### Additional provider/model setup details

_No response_

### Logs, screenshots, and evidence

```shell
15:05:42 [INFO] cron: started, jobs: 0, nextWakeAtMs: null
15:14:40 [ERROR] gateway connect failed: pairing required (CLI cron list)
15:16:33 [ERROR] gateway connect failed: pairing required (CLI cron list) 
CLI 卡住原因： Gateway 绑定 LAN 地址 + dangerouslyDisableDeviceAuth=true + pairing 模式冲突。
```

### Impact and severity

_No response_

### Additional information

之前版本能用 为openclaw 2026.3.28   升级到openclaw4.9就出现问题  回退没问题 升级版本就出现的

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ Docs: https://docs.openclaw.ai
 ### Fixes
 
 - Google Chat: preserve reply text when a typing indicator message is deleted or can no longer be updated, so media captions and first text chunks are resent instead of silently disappearing. (#71498) Thanks @colin-lgtm.
+- Cron: tolerate malformed legacy job rows in startup, main-session system-event payloads, and human-readable `cron list` output so missing `state`, `payload.text`, or display fields no longer crash the scheduler or CLI. Fixes #66016, #65916, #64137, #57872, #59968, #63813, #52804, and #43163.
 - Heartbeat: clamp oversized scheduler delays through the shared safe timer helper, preventing `every` values over Node's timeout cap from becoming a 1 ms crash loop. Fixes #71414. (#71478) Thanks @hclsys.
 - Control UI/chat: collapse assistant token/model context details behind an explicit Context disclosure and show full dates in message footers, making historical transcript timing clear without noisy default metadata. (#71337) Thanks @BunsDev.
 - OpenAI/Codex OAuth: explain `unsupported_country_region_territory` token-exchange failures with a proxy/region hint instead of surfacing a generic OAuth error. Fixes #51175.

--- a/src/cli/cron-cli/shared.test.ts
+++ b/src/cli/cron-cli/shared.test.ts
@@ -78,6 +78,22 @@ describe("printCronList", () => {
     expect(logs.some((line) => line.includes("isolated"))).toBe(true);
   });
 
+  it("tolerates malformed rows in human-readable output", () => {
+    const { logs, runtime } = createRuntimeLogCapture();
+    const malformedJob = {
+      id: "malformed-job",
+      name: undefined,
+      enabled: true,
+      sessionTarget: undefined,
+      payload: undefined,
+      schedule: undefined,
+      state: undefined,
+    } as unknown as CronJob;
+
+    expect(() => printCronList([malformedJob], runtime)).not.toThrow();
+    expect(logs.some((line) => line.includes("malformed-job"))).toBe(true);
+  });
+
   it("shows stagger label for cron schedules", () => {
     const { logs, runtime } = createRuntimeLogCapture();
     const job = createBaseJob({

--- a/src/cli/cron-cli/shared.ts
+++ b/src/cli/cron-cli/shared.ts
@@ -156,7 +156,17 @@ const CRON_DELIVERY_PAD = 64;
 const CRON_AGENT_PAD = 10;
 const CRON_MODEL_PAD = 20;
 
-const pad = (value: string, width: number) => value.padEnd(width);
+const stringifyCell = (value: unknown, fallback = "-") => {
+  if (typeof value === "string") {
+    return value;
+  }
+  if (typeof value === "number" || typeof value === "boolean") {
+    return String(value);
+  }
+  return fallback;
+};
+
+const pad = (value: unknown, width: number) => stringifyCell(value).padEnd(width);
 
 const truncate = (value: string, width: number) => {
   if (value.length <= width) {
@@ -200,12 +210,15 @@ const formatRelative = (ms: number | null | undefined, nowMs: number) => {
   return delta >= 0 ? `in ${label}` : `${label} ago`;
 };
 
-const formatSchedule = (schedule: CronSchedule) => {
-  if (schedule.kind === "at") {
+const formatSchedule = (schedule: CronSchedule | undefined) => {
+  if (schedule?.kind === "at") {
     return `at ${formatIsoMinute(schedule.at)}`;
   }
-  if (schedule.kind === "every") {
+  if (schedule?.kind === "every") {
     return `every ${formatDurationHuman(schedule.everyMs)}`;
+  }
+  if (schedule?.kind !== "cron") {
+    return "-";
   }
   const base = schedule.tz ? `cron ${schedule.expr} @ ${schedule.tz}` : `cron ${schedule.expr}`;
   const staggerMs = resolveCronStaggerMs(schedule);
@@ -219,10 +232,11 @@ const formatStatus = (job: CronJob) => {
   if (!job.enabled) {
     return "disabled";
   }
-  if (job.state.runningAtMs) {
+  const state = job.state ?? {};
+  if (state.runningAtMs) {
     return "running";
   }
-  return job.state.lastStatus ?? "idle";
+  return state.lastStatus ?? "idle";
 };
 
 export function coerceCronDeliveryPreviews(value: unknown): Map<string, CronDeliveryPreview> {
@@ -275,17 +289,18 @@ export function printCronList(
   const now = Date.now();
 
   for (const job of jobs) {
+    const state = job.state ?? {};
     const idLabel = pad(job.id, CRON_ID_PAD);
-    const nameLabel = pad(truncate(job.name, CRON_NAME_PAD), CRON_NAME_PAD);
+    const nameLabel = pad(truncate(stringifyCell(job.name), CRON_NAME_PAD), CRON_NAME_PAD);
     const scheduleLabel = pad(
       truncate(formatSchedule(job.schedule), CRON_SCHEDULE_PAD),
       CRON_SCHEDULE_PAD,
     );
     const nextLabel = pad(
-      job.enabled ? formatRelative(job.state.nextRunAtMs, now) : "-",
+      job.enabled ? formatRelative(state.nextRunAtMs, now) : "-",
       CRON_NEXT_PAD,
     );
-    const lastLabel = pad(formatRelative(job.state.lastRunAtMs, now), CRON_LAST_PAD);
+    const lastLabel = pad(formatRelative(state.lastRunAtMs, now), CRON_LAST_PAD);
     const statusRaw = formatStatus(job);
     const statusLabel = pad(statusRaw, CRON_STATUS_PAD);
     const targetLabel = pad(job.sessionTarget ?? "-", CRON_TARGET_PAD);
@@ -297,7 +312,7 @@ export function printCronList(
     const agentLabel = pad(truncate(job.agentId ?? "-", CRON_AGENT_PAD), CRON_AGENT_PAD);
     const modelLabel = pad(
       truncate(
-        (job.payload.kind === "agentTurn" ? job.payload.model : undefined) ?? "-",
+        (job.payload?.kind === "agentTurn" ? job.payload.model : undefined) ?? "-",
         CRON_MODEL_PAD,
       ),
       CRON_MODEL_PAD,
@@ -339,7 +354,7 @@ export function printCronList(
         ? colorize(rich, theme.info, deliveryLabel)
         : colorize(rich, theme.muted, deliveryLabel),
       coloredAgent,
-      job.payload.kind === "agentTurn" && job.payload.model
+      job.payload?.kind === "agentTurn" && job.payload.model
         ? colorize(rich, theme.info, modelLabel)
         : colorize(rich, theme.muted, modelLabel),
     ].join(" ");

--- a/src/cron/service.jobs.test.ts
+++ b/src/cron/service.jobs.test.ts
@@ -1,5 +1,10 @@
 import { describe, expect, it } from "vitest";
-import { applyJobPatch, createJob, recomputeNextRuns } from "./service/jobs.js";
+import {
+  applyJobPatch,
+  createJob,
+  recomputeNextRuns,
+  resolveJobPayloadTextForMain,
+} from "./service/jobs.js";
 import type { CronServiceState } from "./service/state.js";
 import { DEFAULT_TOP_OF_HOUR_STAGGER_MS } from "./stagger.js";
 import type { CronJob, CronJobPatch } from "./types.js";
@@ -682,6 +687,20 @@ describe("createJob delivery defaults", () => {
       payload: { kind: "systemEvent", text: "ping" },
     });
     expect(job.delivery).toBeUndefined();
+  });
+
+  it("uses legacy systemEvent message text without throwing", () => {
+    const state = createMockState(now, { defaultAgentId: "main" });
+    const job = createJob(state, {
+      name: "legacy system event",
+      enabled: true,
+      schedule: { kind: "every", everyMs: 60_000 },
+      sessionTarget: "main",
+      wakeMode: "now",
+      payload: { kind: "systemEvent", message: "legacy text" } as never,
+    });
+
+    expect(resolveJobPayloadTextForMain(job)).toBe("legacy text");
   });
 });
 

--- a/src/cron/service/normalize.ts
+++ b/src/cron/service/normalize.ts
@@ -63,7 +63,12 @@ export function inferLegacyName(job: {
 
 export function normalizePayloadToSystemText(payload: CronPayload) {
   if (payload.kind === "systemEvent") {
-    return payload.text.trim();
+    const text = (payload as { text?: unknown }).text;
+    if (typeof text === "string") {
+      return text.trim();
+    }
+    const legacyMessage = (payload as { message?: unknown }).message;
+    return typeof legacyMessage === "string" ? legacyMessage.trim() : "";
   }
-  return payload.message.trim();
+  return typeof payload.message === "string" ? payload.message.trim() : "";
 }

--- a/src/cron/service/ops.regression.test.ts
+++ b/src/cron/service/ops.regression.test.ts
@@ -16,7 +16,7 @@ import {
   waitForActiveTasks,
 } from "../../process/command-queue.js";
 import { CommandLane } from "../../process/lanes.js";
-import { enqueueRun, run } from "./ops.js";
+import { enqueueRun, run, start } from "./ops.js";
 import type { CronEvent } from "./state.js";
 import { createCronServiceState } from "./state.js";
 import { onTimer } from "./timer.js";
@@ -27,6 +27,37 @@ const opsRegressionFixtures = setupCronRegressionFixtures({
 });
 
 describe("cron service ops regressions", () => {
+  it("does not crash startup when a loaded job is missing state", async () => {
+    const scheduledAt = Date.now() + 60_000;
+    const store = opsRegressionFixtures.makeStorePath();
+    const state = createCronServiceState({
+      cronEnabled: true,
+      storePath: store.storePath,
+      log: noopLogger,
+      enqueueSystemEvent: vi.fn(),
+      requestHeartbeatNow: vi.fn(),
+      runIsolatedAgentJob: vi.fn(),
+    });
+    state.store = {
+      version: 1,
+      jobs: [
+        {
+          ...createIsolatedRegressionJob({
+            id: "missing-state-startup",
+            name: "missing-state-startup",
+            scheduledAt,
+            schedule: { kind: "at", at: new Date(scheduledAt).toISOString() },
+            payload: { kind: "agentTurn", message: "noop" },
+          }),
+          state: undefined as never,
+        },
+      ],
+    };
+
+    await expect(start(state)).resolves.toBeUndefined();
+    expect(state.store.jobs[0]?.state).toEqual(expect.any(Object));
+  });
+
   it("skips forced manual runs while a timer-triggered run is in progress", async () => {
     const store = opsRegressionFixtures.makeStorePath();
     const dueAt = Date.now() - 1;

--- a/src/cron/service/ops.ts
+++ b/src/cron/service/ops.ts
@@ -96,6 +96,7 @@ export async function start(state: CronServiceState) {
     await ensureLoaded(state, { skipRecompute: true });
     const jobs = state.store?.jobs ?? [];
     for (const job of jobs) {
+      job.state ??= {};
       if (typeof job.state.runningAtMs === "number") {
         state.deps.log.warn(
           { jobId: job.id, runningAtMs: job.state.runningAtMs },

